Remove redundant sorting

devcrafter opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove redundant sorting, for example, sorting related ORDER BY clauses in subqueries. Implemented on top of query plan. It does similar optimization as `optimize_duplicate_order_by_and_distinct` regarding `ORDER BY` clauses, but more generic, since it's applied to any redundant sorting steps (not only caused by ORDER BY clause) and applied to subqueries of any depth. Related to #42648

devcrafter committed:
save

devcrafter committed:
Something working

devcrafter committed:
Some tests + setting

devcrafter committed:
Add GROUP BY w/o aggrecation functions analyze + test

devcrafter committed:
Fix type

devcrafter committed:
Fix + CROSS JOIN tests

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
GROUP BY with check if aggr func depends on order

+ tests

devcrafter committed:
Respect stateful functions in the optimization

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Polishing

devcrafter commented on src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp:
```diff
@@ -0,0 +1,190 @@
+    return typeid(*current_step).name();
+}
+
+void printStepName(const char * prefix, const QueryPlan::Node * node)
```
Remove debugging code

devcrafter commented on src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp:
```diff
@@ -0,0 +1,190 @@
+    return false;
+}
+
+void tryRemoveRedundantOrderBy(QueryPlan::Node * root)
```
I guess, it'd be nicer to implement it as a class like visitor with corresponding methods for top-down/bottom-up traversing

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix: remove ORDER BY before "mergeExpressions" optimization

When removing ORDER BY we remove corresponding expression,
so to remove only ORDER BY expression, do it before "mergeExpressions"
optimization is applied

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Merge branch 'master' into igor/remove_redundant_order_by

devcrafter committed:
Fix window function case

devcrafter committed:
Query plan visitor with debug logs

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix for distributed query

- do not apply optimization in case of UnionStep on top of Sorting

01104_distributed_one_test.sql

devcrafter committed:
Move checks to functions

devcrafter committed:
Fix for distributed queries

+ 01952_optimize_distributed_group_by_sharding_key

devcrafter committed:
Update test to make it stable

+ 02377_optimize_sorting_by_input_stream_properties_explain.sh

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix for full sorting merge join

devcrafter committed:
Avoid to remove sorting expression for now

devcrafter committed:
Stabilizing tests

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Temporary fix for test

devcrafter committed:
Disable the optimization for distributed queries

+ based on ReadFromRemote
- improve by disabling optimization only between Union and ReadFromRemote

devcrafter committed:
Fix distributed query check

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix: distributed query with cluster() table function

devcrafter committed:
Revert mistakenly changed test reference

devcrafter committed:
Remove current sorting only if next is full sorting

devcrafter committed:
If aggregation step is in order, - do not remove underneath sorting step

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter commented:
MSan report - https://github.com/ClickHouse/ClickHouse/issues/44494

devcrafter committed:
Enable the optimization back for a test

02404_memory_bound_merging.sql

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Remove incorrect assert

ExpressionStep is not always under SortingStep
It can be ReadFromRemote in distributed query

devcrafter committed:
Remove incorrect assert

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp:
```diff
@@ -0,0 +1,317 @@
+
+    static bool visitTopDown(QueryPlan::Node *, QueryPlan::Node *) { return true; }
+
+    void visitBottomUp(QueryPlan::Node * current_node, QueryPlan::Node *)
```
I have a warning `Function 'visitBottomUp' hides a non-virtual function from class 'QueryPlanVisitor<IsQueryDistributed, 1>'`. It's not very important, but we can consider renaming in QueryPlanVisitor

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp:
```diff
@@ -0,0 +1,317 @@
+            const auto * expr = typeid_cast<const ExpressionStep *>(step);
+            if (expr)
+            {
+                if (expr->getExpression()->hasStatefulFunctions())
```
Probably we need the same chack for FilterStep

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp:
```diff
@@ -0,0 +1,317 @@
+
+class RemoveRedundantOrderBy : public QueryPlanVisitor<RemoveRedundantOrderBy, debug_logging_enabled>
+{
+    std::vector<QueryPlan::Node *> nodes_affect_order;
```
Comment would be helpful.

Maybe not very important, but I'd better try to store the last node from `nodes_affect_order` in the stack frame.
For this, you can specify a template structure that you can store in the stack.

Also, ideally, somehow make a stack private for derived classes.

KochetovNicolai reviewed (approved):
Generally ok (if we add FilterStep to check)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix review comments

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Fix typo

devcrafter committed:
Disable diagnostics

+ cleanup
+ disable optimization in sort performance test since it removes sorting
  at all

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter committed:
Test for aggr func w/o group by

devcrafter committed:
Rename settings to remove_redundant_sorting

devcrafter committed:
Renaming according to setting

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter commented:
[Fatal messages in stress test](https://s3.amazonaws.com/clickhouse-test-reports/43905/042f61ab92264daecccef6c2235c14b4f746fc7e/stress_test__debug_/fatal_messages.txt), [CI link](https://s3.amazonaws.com/clickhouse-test-reports/43905/042f61ab92264daecccef6c2235c14b4f746fc7e/stress_test__debug_.html) :
```
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:46:31.403884 [ 31178 ] {a9875e83-3cf9-413d-826f-9c3d975a2bc5} <Fatal> : Logical error: 'Join is supported only for pipelines with one output port'.
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.413142 [ 47753 ] {} <Fatal> BaseDaemon: ########################################
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.413953 [ 47753 ] {} <Fatal> BaseDaemon: (version 22.13.1.1, build id: 174BAE212B423703828ACC2B98068E63786671A6) (from thread 31178) (query_id: a9875e83-3cf9-413d-826f-9c3d975a2bc5) (query: SELECT bitmapCardinality(day_today) AS today_users, bitmapCardinality(day_before) AS before_users, bitmapOrCardinality(day_today, day_before) AS all_users, bitmapAndCardinality(day_today, day_before) AS old_users, bitmapAndnotCardinality(day_today, day_before) AS new_users, bitmapXorCardinality(day_today, day_before) AS diff_users FROM ( SELECT city_id, groupBitmapState( uid ) AS day_today FROM bitmap_test WHERE pickup_date = '2019-01-02' GROUP BY city_id ORDER BY city_id ) js1 ALL LEFT JOIN ( SELECT city_id, groupBitmapState( uid ) AS day_before FROM bitmap_test WHERE pickup_date = '2019-01-01' GROUP BY city_id ORDER BY city_id ) js2 USING city_id;) Received signal Aborted (6)
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.414340 [ 47753 ] {} <Fatal> BaseDaemon: 
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.414713 [ 47753 ] {} <Fatal> BaseDaemon: Stack trace: 0x7fbed1c6c00b 0x7fbed1c4b859 0x21200656 0x21200715 0x21200865 0x187e562a 0x2819856f 0x2ba8c9bc 0x2baa05df 0x295d4d71 0x29a6f0ea 0x29a6afc4 0x2b38c9b4 0x2b39d185 0x2ff577d9 0x2ff5801c 0x301a7254 0x301a3ffa 0x301a2dde 0x7fbed1e23609 0x7fbed1d48133
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.415120 [ 47753 ] {} <Fatal> BaseDaemon: 4. raise @ 0x7fbed1c6c00b in ?
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.415360 [ 47753 ] {} <Fatal> BaseDaemon: 5. abort @ 0x7fbed1c4b859 in ?
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.821774 [ 47753 ] {} <Fatal> BaseDaemon: 6. /build/build_docker/../src/Common/Exception.cpp:41: DB::abortOnFailedAssertion(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) @ 0x21200656 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:56.941782 [ 47753 ] {} <Fatal> BaseDaemon: 7. /build/build_docker/../src/Common/Exception.cpp:64: DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*>> const&) @ 0x21200715 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:57.043891 [ 47753 ] {} <Fatal> BaseDaemon: 8. /build/build_docker/../src/Common/Exception.cpp:78: DB::Exception::Exception(DB::Exception::MessageMasked const&, int, bool) @ 0x21200865 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:57.167260 [ 47753 ] {} <Fatal> BaseDaemon: 9. /build/build_docker/../src/Common/Exception.h:41: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool) @ 0x187e562a in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:57.790904 [ 47753 ] {} <Fatal> BaseDaemon: 10. /build/build_docker/../src/QueryPipeline/QueryPipelineBuilder.cpp:354: DB::QueryPipelineBuilder::joinPipelinesYShaped(std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::shared_ptr<DB::IJoin>, DB::Block const&, unsigned long, std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>*) @ 0x2819856f in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:58.117524 [ 47753 ] {} <Fatal> BaseDaemon: 11. /build/build_docker/../src/Processors/QueryPlan/JoinStep.cpp:38: DB::JoinStep::updatePipeline(std::__1::vector<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::allocator<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&) @ 0x2ba8c9bc in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:58.354250 [ 47753 ] {} <Fatal> BaseDaemon: 12. /build/build_docker/../src/Processors/QueryPlan/QueryPlan.cpp:187: DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&) @ 0x2baa05df in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:58.997904 [ 47753 ] {} <Fatal> BaseDaemon: 13. /build/build_docker/../src/Interpreters/InterpreterSelectWithUnionQuery.cpp:380: DB::InterpreterSelectWithUnionQuery::execute() @ 0x295d4d71 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:47:59.748408 [ 47753 ] {} <Fatal> BaseDaemon: 14. /build/build_docker/../src/Interpreters/executeQuery.cpp:705: DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x29a6f0ea in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:00.053072 [ 47753 ] {} <Fatal> BaseDaemon: 15. /build/build_docker/../src/Interpreters/executeQuery.cpp:1104: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x29a6afc4 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:00.693271 [ 47753 ] {} <Fatal> BaseDaemon: 16. /build/build_docker/../src/Server/TCPHandler.cpp:378: DB::TCPHandler::runImpl() @ 0x2b38c9b4 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.184866 [ 47753 ] {} <Fatal> BaseDaemon: 17. /build/build_docker/../src/Server/TCPHandler.cpp:1931: DB::TCPHandler::run() @ 0x2b39d185 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.220573 [ 47753 ] {} <Fatal> BaseDaemon: 18. /build/build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x2ff577d9 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.330324 [ 47753 ] {} <Fatal> BaseDaemon: 19. /build/build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x2ff5801c in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.430185 [ 47753 ] {} <Fatal> BaseDaemon: 20. /build/build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199: Poco::PooledThread::run() @ 0x301a7254 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.501286 [ 47753 ] {} <Fatal> BaseDaemon: 21. /build/build_docker/../contrib/poco/Foundation/src/Thread.cpp:56: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x301a3ffa in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.551658 [ 47753 ] {} <Fatal> BaseDaemon: 22. /build/build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345: Poco::ThreadImpl::runnableEntry(void*) @ 0x301a2dde in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.562212 [ 47753 ] {} <Fatal> BaseDaemon: 23. ? @ 0x7fbed1e23609 in ?
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.562543 [ 47753 ] {} <Fatal> BaseDaemon: 24. __clone @ 0x7fbed1d48133 in ?
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:01.562781 [ 47753 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
/var/log/clickhouse-server/clickhouse-server.err.log:2023.01.06 21:48:46.742502 [ 1568 ] {} <Fatal> Application: Child process was terminated by signal 6.
```
cc @vdimir Is it something known?

vdimir commented:
Checked on commit `042f61ab92264daecccef6c2235c14b4f746fc7e` from this PR

Repro:


```sql
DROP TABLE IF EXISTS t;
CREATE TABLE t ( city_id UInt32 ) ENGINE = Memory;
INSERT INTO t VALUES (1), (2);

SET max_rows_in_set_to_optimize_join = 0; --- not necessary

SET join_algorithm = 'full_sorting_merge';
SET query_plan_remove_redundant_order_by = 1;

explain plan description = 1, sorting = 1
SELECT *
FROM ( SELECT city_id FROM t GROUP BY city_id ORDER BY city_id ) js1
JOIN ( SELECT city_id FROM t GROUP BY city_id ORDER BY city_id ) js2
USING city_id;
SET query_plan_remove_redundant_order_by = 0;
SELECT '----';

explain plan description = 1, sorting = 1
SELECT *
FROM ( SELECT city_id FROM t GROUP BY city_id ORDER BY city_id ) js1
JOIN ( SELECT city_id FROM t GROUP BY city_id ORDER BY city_id ) js2
USING city_id;
```

But actually streams before join is sorted:

```
Expression ((Projection + Before ORDER BY))
Sorting (None)
  Join (JOIN YShaped)
  Sorting (None)
    Sorting (Sort Left before JOIN)
    Sorting (Global): city_id ASC
      Expression ((Before JOIN + (Projection + Before ORDER BY)))
      Sorting (None)
        Aggregating
        Sorting (None)
          Expression (Before GROUP BY)
          Sorting (None)
            ReadFromStorage (Memory)
            Sorting (None)
    Sorting (Sort Right before JOIN)
    Sorting (Global): js2.city_id ASC
      Expression ((Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))
      Sorting (None)
        Aggregating
        Sorting (None)
          Expression (Before GROUP BY)
          Sorting (None)
            ReadFromStorage (Memory)
            Sorting (None)
----
Expression ((Projection + Before ORDER BY))
Sorting (None)
  Join (JOIN YShaped)
  Sorting (None)
    Sorting (Sort Left before JOIN)
    Sorting (Global): city_id ASC
      Expression ((Before JOIN + Projection))
      Sorting (Global): city_id ASC
        Sorting (Sorting for ORDER BY)
        Sorting (Global): city_id ASC
          Expression (Before ORDER BY)
          Sorting (None)
            Aggregating
            Sorting (None)
              Expression (Before GROUP BY)
              Sorting (None)
                ReadFromStorage (Memory)
                Sorting (None)
    Sorting (Sort Right before JOIN)
    Sorting (Global): js2.city_id ASC
      Expression ((Joined actions + (Rename joined columns + Projection)))
      Sorting (Global): city_id ASC
        Sorting (Sorting for ORDER BY)
        Sorting (Global): city_id ASC
          Expression (Before ORDER BY)
          Sorting (None)
            Aggregating
            Sorting (None)
              Expression (Before GROUP BY)
              Sorting (None)
                ReadFromStorage (Memory)
                Sorting (None)
```


Need to add some additional info to error message to check what `left` and `right` actually is

https://github.com/ClickHouse/ClickHouse/blob/119501f1d9e33653e2cd5de6c82ed8a54f860581/src/QueryPipeline/QueryPipelineBuilder.cpp#L353-L354

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

vdimir commented:
@devcrafter I've pushed commit with fix for the case above, feel free to drop it if you'll find proper solution.

devcrafter committed:
Fix: update input/output stream properties

After removing sorting step we need to update sorting properties of
input/ouput streams

devcrafter committed:
Rename source file according to implementation

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/remove_redundant_order_by

devcrafter commented:
> @devcrafter I've pushed commit with fix for the case above, feel free to drop it if you'll find proper solution.

@vdimir Thanks! The proper fix is [here](https://github.com/ClickHouse/ClickHouse/pull/43905/commits/6328e02f22f1f46fa25f97a672a8ee74762ed632). Your commit contains better checks during building pipeline for JOIN, so I think, it'd be nice to bring them in but by other PR.

devcrafter committed:
Merge branch 'master' into igor/remove_redundant_order_by

devcrafter commented:
Stress test (ubsan) failure related to https://github.com/ClickHouse/ClickHouse/issues/44912
